com.unity.render-pipelines.core/Runtime/XR/XRSystem.cs

```diff
@@ -0,0 +1,372 @@
+        {
+            RefreshDeviceInfo();
+
+            if (s_Layout.GetActivePasses().Count > 0)
+            {
+                Debug.LogWarning("Render Pipeline error : XRSystem.EndLayout() was not called!");
+                EndLayout();
+            }
+
+            return s_Layout;
+        }
+
+        /// <summary>
+        /// Used by the render pipeline to complete the XR layout at the end of the frame.
+        /// </summary>
+        public static void EndLayout()
+        {
+            if (dumpDebugInfo)
+                s_Layout.LogDebugInfo();
+
+            s_Layout.Clear();
+        }
+
+        /// <summary>
+        /// Used by the render pipeline to render the mirror view to the gameview, as configured by the XR device.
+        /// </summary>
+        /// <param name="cmd"></param>
+        /// <param name="camera"></param>
+        public static void RenderMirrorView(CommandBuffer cmd, Camera camera)
+        {
+        #if ENABLE_VR && ENABLE_XR_MODULE
+            XRMirrorView.RenderMirrorView(cmd, camera, s_MirrorViewMaterial, s_Display);
+        #endif
+        }
+
+        /// <summary>
+        /// Free up resources used by the system.
+        /// </summary>
+        public static void Dispose()
+        {
+            CoreUtils.Destroy(s_OcclusionMeshMaterial);
+            CoreUtils.Destroy(s_MirrorViewMaterial);
+        }
+
+        // Used by the render pipeline to communicate to the XR device the range of the depth buffer.
+        internal static void SetDisplayZRange(float zNear, float zFar)
+        {
+        #if ENABLE_VR && ENABLE_XR_MODULE
+            if (s_Display != null)
+            {
+                s_Display.zNear = zNear;
+                s_Display.zFar = zFar;
+            }
+        #endif
+        }
+
+        internal static void SetDisplaySync()
+        {
+            // Disable vsync on the main display when rendering to a XR device.
+            // XRTODO : expose public property to control this behavior
+            QualitySettings.vSyncCount = 0;
+
+            // On Android and iOS, vSyncCount is ignored and all frame rate control is done using Application.targetFrameRate.
+            if (Application.platform == RuntimePlatform.Android)
+            {
+                float frameRate = 300.0f;
+                //frameRate = s_Display.TryGetDisplayRefreshRate(out float refreshRate) ? refreshRate : frameRate;
+
+                // XRTODO : move out, causing issues with HDRP timing
+                Application.targetFrameRate = Mathf.CeilToInt(frameRate);
+            }
+        }
+
+        // XRTODO : expose as public API
+        static void SetLayoutOverride(Action<XRLayout, Camera> action)
+        {
+            s_LayoutOverride = action;
+        }
+
+        // Disable legacy VR system before rendering first frame
+        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.BeforeSplashScreen)]
+        static void XRSystemInit()
```

Comment:
I am curious - does this work?
I thought that first frame was always builtin since the SRP are not yet initialized. but since we are checking the serialized value it may work?

Reply:
Yes this has been working like this for 2 years now. I don't think we have a better way but I'm open to ideas.